library/init/meta/derive.lean

```diff
@@ -0,0 +1,61 @@
+/-
+Copyright (c) 2017 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Ullrich
+
+Attribute that can automatically derive typeclass instances.
+-/
+prelude
+import init.meta.attribute
+import init.meta.interactive_base
+import init.meta.mk_dec_eq_instance
+import init.meta.mk_has_sizeof_instance
+import init.meta.mk_inhabited_instance
+
+open lean
+open interactive.types
+open tactic
+
+meta def derive_handler := Π (cls : pexpr) (decl : name), tactic bool
+
+@[user_attribute]
+meta def derive_handler_attr : user_attribute :=
+{ name := `derive_handler, descr := "register a definition of type `derive_handler` for use in the [derive] attribute" }
+
+private meta def try_handlers (p : pexpr) (n : name) : list derive_handler → tactic unit
+| []      := fail format!"failed to find a derive handler for '{p}'"
+| (h::hs) :=
+do success ← h p n,
+   when (¬success) $
+     try_handlers hs
+
+@[user_attribute] meta def derive_attr : user_attribute unit (list pexpr) :=
+{ name := `derive, descr := "automatically derive typeclass instances",
+  parser := pexpr_list_or_texpr,
+  after_set := some (λ n _ _,
+  do ps ← derive_attr.get_param n,
```

Comment:
Note the recursive use of `derive_attr` -- unbounded recursion can be helpful indeed.